python/dgl/nn/pytorch/linear.py

```diff
@@ -0,0 +1,178 @@
+"""Various commonly used linear modules"""
+# pylint: disable= no-member, arguments-differ, invalid-name, W0235
+import torch
+import torch.nn as nn
+
+from ...ops import segment_mm, gather_mm
+
+__all__ = ['TypedLinear']
+
+class TypedLinear(nn.Module):
+    r"""Linear transformation according to types.
+
+    For each sample of the input batch :math:`x \in X`, apply linear transformation
+    :math:`xW_t`, where :math:`t` is the type of :math:`x`.
+
+    The module supports two regularization methods (basis-decomposition and
+    block-diagonal-decomposition) proposed by "`Modeling Relational Data
+    with Graph Convolutional Networks <https://arxiv.org/abs/1703.06103>`__"
+
+    The basis regularization decomposes :math:`W_t` by:
+
+    .. math::
+
+       W_t^{(l)} = \sum_{b=1}^B a_{tb}^{(l)}V_b^{(l)}
+
+    where :math:`B` is the number of bases, :math:`V_b^{(l)}` are linearly combined
+    with coefficients :math:`a_{tb}^{(l)}`.
+
+    The block-diagonal-decomposition regularization decomposes :math:`W_t` into :math:`B`
+    number of block diagonal matrices. We refer :math:`B` as the number of bases.
+
+    The block regularization decomposes :math:`W_t` by:
+
+    .. math::
+
+       W_t^{(l)} = \oplus_{b=1}^B Q_{tb}^{(l)}
+
+    where :math:`B` is the number of bases, :math:`Q_{tb}^{(l)}` are block
+    bases with shape :math:`R^{(d^{(l+1)}/B)*(d^{l}/B)}`.
+
+    Parameters
+    ----------
+    in_size : int
+        Input feature size.
+    out_size : int
+        Output feature size.
+    num_types : int
+        Total number of types.
+    regularizer : str, optional
+        Which weight regularizer to use "basis" or "bdd":
+
+         - "basis" is short for basis-decomposition.
+         - "bdd" is short for block-diagonal-decomposition.
+
+        Default applies no regularization.
+    num_bases : int, optional
+        Number of bases. Needed when ``regularizer`` is specified. Default: ``None``.
+
+    Examples
+    --------
+
+    No regularization.
+    >>> from dgl.nn import TypedLinear
+    >>> import torch
+    >>>
+    >>> x = torch.randn(100, 32)
+    >>> x_type = torch.randint(0, 5, (100,))
+    >>> m = TypedLinear(32, 64, 5)
+    >>> y = m(x, x_type)
+    >>> print(y.shape)
+    torch.Size([100, 64])
+
+    With basis regularization
+    >>> from dgl.nn import TypedLinear
+    >>> import torch
+    >>>
+    >>> x = torch.randn(100, 32)
+    >>> x_type = torch.randint(0, 5, (100,))
+    >>> m = TypedLinear(32, 64, 5, regularizer='basis', num_bases=4)
+    >>> y = m(x, x_type)
+    >>> print(y.shape)
+    torch.Size([100, 64])
+    """
+    def __init__(self, in_size, out_size, num_types,
```

Comment:
Can we make `in_size` optionally a list/dict, which indicates the input sizes for each type enumerated by `num_types`?  I understand that this would then make the regularization methods not work.  Whether or not this makes sense depends on the point of this module.  If it's only to encapsulate this typed linear transform that R-GCN does, then it perhaps doesn't make sense to do as I suggest.  However, if this is R-GCN specific, then it raises the question of why we should separate out this component.

Alternatively, users might want to use this as a general PyTorch convenience module for projecting features of different node types into the same embedding space.  In that case, we do not need to support those regularization methods here, and instead can have those as stand-alone modules that appropriately assume everything is of the same dimensionality.

To summarize, I think it might make more sense to remove these regularization methods from this module and instead make this more general, in which node types with different feature lengths can be handled, and the purpose is to project into a common dimensionality.  Otherwise I don't understand where this module would be used other than in R-GCN, which begs the question as to why it's kept separate.

Reply:
> Can we make in_size optionally a list/dict,
> Alternatively, users might want to use this as a general PyTorch convenience module for projecting features of different node types into the same embedding space. In that case, we do not need to support those regularization methods here, and instead can have those as stand-alone modules that appropriately assume everything is of the same dimensionality.

We decided to have a separate module for this. See [`HeteroLinear`](https://docs.dgl.ai/en/latest/api/python/nn.pytorch.html#heterolinear). `TypedLinear` and `HeteroLinear` are quite different in many aspects, i.e., `TypedLinear` only allows embeddings of the same length but supports regularization while `HeteroLinear` supports embeddings of different lengths stored in a dict but no weight regularization. Separating the two implementations also makes the code much shorter (saving quite a lot of if-else).

> Otherwise I don't understand where this module would be used other than in R-GCN, which begs the question as to why it's kept separate.

It is also used in `HGTConv`. See the code [here](https://github.com/dmlc/dgl/blob/49563d28d1b1279c9f7f5c2ba3cd9d9f9dbf33eb/python/dgl/nn/pytorch/conv/hgtconv.py).